### PR TITLE
Broadcast join if build estimation is small and from HBO

### PR DESCRIPTION
**Pull request by feilong-liu**
## Description
Previously I added this session property to do broadcast join when build side is small and probe side is unknown. However, in production we found that it will cause problem when the build side estimation is wrong. However, HBO is a confident source and we should be able to enable it when the estimation is from HBO.

## Motivation and Context
To improve performance for queries with unknown probe and small build.

## Impact
To improve performance for queries with unknown probe and small build.

## Test Plan
It's an easy change, and also tested locally end to end.

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Improve session property ``property-use_broadcast_when_buildsize_small_probeside_unknown`` to do broadcast join when probe side size is unknown and build side estimation from HBO is small
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineSemiJoinDistributionType.java [hunk 1]
```diff
@@ -124,6 +128,12 @@ private PlanNode getCostBasedDistributionType(SemiJoinNode node, Context context
         possibleJoinNodes.add(getSemiJoinNodeWithCost(node.withDistributionType(PARTITIONED), context));
 
         if (possibleJoinNodes.stream().anyMatch(result -> result.getCost().hasUnknownComponents())) {
+            if (isUseBroadcastJoinWhenBuildSizeSmallProbeSizeUnknownEnabled(context.getSession()) && possibleJoinNodes.stream().anyMatch(result -> ((SemiJoinNode) result.getPlanNode()).getDistributionType().get().equals(REPLICATED))) {
```
Comment (feilong-liu) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineSemiJoinDistributionType.java:131]: Also enable it for semi join

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 2]
```diff
@@ -126,7 +127,10 @@ private PlanNode getCostBasedJoin(JoinNode joinNode, Context context)
         if (possibleJoinNodes.stream().anyMatch(result -> result.getCost().hasUnknownComponents()) || possibleJoinNodes.isEmpty()) {
             // TODO: currently this session parameter is added so as to roll out the plan change gradually, after proved to be a better choice, make it default and get rid of the session parameter here.
             if (isUseBroadcastJoinWhenBuildSizeSmallProbeSizeUnknownEnabled(context.getSession()) && possibleJoinNodes.stream().anyMatch(result -> ((JoinNode) result.getPlanNode()).getDistributionType().get().equals(REPLICATED))) {
-                return getOnlyElement(possibleJoinNodes.stream().filter(result -> ((JoinNode) result.getPlanNode()).getDistributionType().get().equals(REPLICATED)).map(x -> x.getPlanNode()).collect(toImmutableList()));
+                JoinNode broadcastJoin = (JoinNode) getOnlyElement(possibleJoinNodes.stream().filter(result -> ((JoinNode) result.getPlanNode()).getDistributionType().get().equals(REPLICATED)).map(x -> x.getPlanNode()).collect(toImmutableList()));
```
Comment (kaikalur) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:130]: We should do it always - we trust history and if HBO says build side is small, we just broadcast it.

**Comment by jaystarshot** [comment]
This might be a good time to fix and use the correct confidence level in planNodeStatsEstimate

**Comment by jaystarshot** [comment]
But that might involve additional testing and could be an overkill for this case so will leave you to decide

**Comment by steveburnett** [comment]
Suggest revising the release note entry to follow the [Order of changes](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#order-of-changes) in the Release Notes Guidelines. Maybe something like this? Edit this as needed, please.

```
== RELEASE NOTES ==

General Changes
* Improve session property ``property-name`` (to?) broadcast join when probe side size is unknown and build side estimation from HBO is small
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 2 above)
Comment (ClarenceThreepwood) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:130]: nit: limit lines to [180 chars](https://github.com/prestodb/presto/blob/master/CONTRIBUTING.md#code-style:~:text=Lines%20should%20be%20no%20more%20than%20180%20chars.%20Intellij%20%E2%80%9CReformat%20code%E2%80%9D%20does%20NOT%20enforce%20the%20line%20width%2C%20so%20you%20must%20adjust%20it%20yourself.)

**Review by ClarenceThreepwood — COMMENTED** [review]
I have a couple of related questions

1. It is well known that incorrect cost estimates can negatively influence optimizer planning. With this change, if the cost were correct, then we lose the benefit of the optimization - do we not? 

2. I agree that the HBO is a better source of truth. I wonder if we should introduce a config parameter (or a set of configs) that govern whether some optimizations should apply based on stats sources? Maintenance becomes hard if we pepper the code-base with hard-coded special cases for HBO optimizations.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java (hunk 2 above)
Comment (feilong-liu) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:130]: This depends on whether the other side of the join is small or not, i.e. if both sides have estimations showing data size is small, whether we should do partitioned join or broadcast join. I think we can open a separate PR for this.

**Comment by feilong-liu** [comment]
> But that might involve additional testing and could be an overkill for this case so will leave you to decide

Agree that we need to have different confidence level for estimations. I will consider this PR as a simple fix for existing features and not including other major changes.

**Comment by feilong-liu** [comment]
> I have a couple of related questions
> 
> 1. It is well known that incorrect cost estimates can negatively influence optimizer planning. With this change, if the cost were correct, then we lose the benefit of the optimization - do we not?

Yes, we will. But fail queries which ran successfully is a larger pain than failing to optimize queries already running successful/failing. And this is the reason why we still cannot turn on this feature internally.

> 2. I agree that the HBO is a better source of truth. I wonder if we should introduce a config parameter (or a set of configs) that govern whether some optimizations should apply based on stats sources? Maintenance becomes hard if we pepper the code-base with hard-coded special cases for HBO optimizations.

I think this is a good idea and not only applicable to this optimization, but also other optimizations. But it also involves larger changes. I think we can open an issue for this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java [hunk 3]
```diff
@@ -126,7 +127,10 @@ private PlanNode getCostBasedJoin(JoinNode joinNode, Context context)
...
+                if (context.getStatsProvider().getStats(broadcastJoin.getBuild()).getSourceInfo() instanceof HistoryBasedSourceInfo) {
+                    return broadcastJoin;
```
Comment (jaystarshot) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/DetermineJoinDistributionType.java:132]: I think, in future whether the previous join was broadcast/hash and if it was efficient/or if it failed etc could also be stored using HBO

**Review by jaystarshot — APPROVED** [review]

**Pull request merged**